app/src/main/java/org/wikipedia/search/SearchResultsFragment.java

```diff
@@ -578,5 +650,88 @@ private Callback callback() {
     private String getSearchLanguageCode() {
         return ((SearchFragment) getParentFragment()).getSearchLanguageCode();
     }
+
+    private void gatherSearchResultsFromAppSources() {
+        addSearchResultFromReadingLists();
+        addSearchResultFromOpenTabs();
+        addSearchResultFromHistoryEntries();
+    }
+
+    private void addSearchResultFromHistoryEntries() {
+        if (loaderManager == null) {
+            loaderManager = LoaderManager.getInstance(requireActivity()).initLoader(SEARCH_FRAGMENT_HISTORY_LOADER_ID, null, loaderCallback);
+        } else {
+            LoaderManager.getInstance(requireActivity()).restartLoader(SEARCH_FRAGMENT_HISTORY_LOADER_ID, null, loaderCallback);
+        }
+    }
+
+    private void addSearchResultFromReadingLists() {
+        List<SearchResult> searchResultsFromAppSources = new ArrayList<>();
+        List<Object> list = ReadingListBehaviorsUtil.INSTANCE.applySearchQuery(currentSearchTerm, ReadingListDbHelper.instance().getAllLists());
+        for (Object o : list) {
+            if (o instanceof ReadingListPage) {
+                ReadingListPage page = (ReadingListPage) o;
+                PageTitle pageTitle = new PageTitle(page.title(), page.wiki(), page.thumbUrl());
+                SearchResult searchResult = new SearchResult(pageTitle);
+                searchResult.setSearchResultTypeWithPriority(READING_LIST_SEARCH_RESULT);
+                searchResultsFromAppSources.add(searchResult);
+                displayResults(searchResultsFromAppSources);
+                break;
+            }
+        }
+    }
+
+    private void addSearchResultFromOpenTabs() {
+        List<SearchResult> searchResultsFromAppSources = new ArrayList<>();
+        List<Tab> tabList = WikipediaApp.getInstance().getTabList();
+        for (Tab tab : tabList) {
+            if (tab.getBackStackPositionTitle() != null && tab.getBackStackPositionTitle().getDisplayText().toLowerCase().contains(currentSearchTerm.toLowerCase())) {
+                SearchResult searchResult = new SearchResult(tab.getBackStackPositionTitle());
+                searchResult.setSearchResultTypeWithPriority(TAB_LIST_SEARCH_RESULT);
+                searchResultsFromAppSources.add(searchResult);
+                displayResults(searchResultsFromAppSources);
+                return;
+            }
+        }
+    }
+
+    private class LoaderCallback implements LoaderManager.LoaderCallbacks<Cursor> {
```

Comment:
Since we're only looking for a single history entry, it's not really necessary to muck around with `CursorLoader`. You can just query the database table directly to find a single result. (And `zip` it into an Rx chain with the other queries)